[1.19] Menus and Screens: An HCI Research Paper (#465)
MinecraftForge/Documentation: docs/gui/menus.md

// In MyMenuExtra, an AbstractContainerMenu subclass
public MyMenuExtra(int containerId, Inventory playerInv, FriendlyByteBuf extraData) {
  super(MY_MENU_EXTRA.get(), containerId);
  // Store extra data from buffer
  // ...
}
```

## `AbstractContainerMenu`

All menus are extended from `AbstractContainerMenu`. A menu takes in two parameters, the [`MenuType`][mt], which represents the type of the menu itself, and the container id, which represents the unique identifier of the menu for the current accessor.

!!! important
    The player can only have 100 unique menus open at once.

Each menu should contain two constructors: one used to initialize the menu on the server and one used to initialize the menu on the client. The constructor used to initialize the menu on the client is the one supplied to the `MenuType`. Any fields that the server menu constructor contains should have some default for the client menu constructor.

```java
// Client menu constructor
public MyMenu(int containerId, Inventory playerInventory) {
  this(containerId, playerInventory);
}

// Server menu constructor
public MyMenu(int containerId, Inventory playerInventory) {
  // ...
}
```

> Each menu implementation must implement two methods: `#stillValid` and [`#quickMoveStack`][qms]. `#stillValid` determines whether the menu should remain open for a given player. This is typically directed to the Forge deprecated `Container#stillValid`. As an alternative, a boolean function which takes in the player will work just as well. The client menu must always return `true` for this method. The vanilla implementation checks whether the player is within eight blocks of where the data storage object is located.

Reviewer: Perhaps "Forge deprecated" should be hyphenated, to signal more clearly that it is a method which Forge deprecates, rather than a Forge method which is also deprecated?
```suggestion
Each menu implementation must implement two methods: `#stillValid` and [`#quickMoveStack`][qms]. `#stillValid` determines whether the menu should remain open for a given player. This is typically directed to the Forge-deprecated `Container#stillValid`. As an alternative, a boolean function which takes in the player will work just as well. The client menu must always return `true` for this method. The vanilla implementation checks whether the player is within eight blocks of where the data storage object is located.
```

```java
// Client menu constructor
public MyMenuAccess(int containerId, Inventory playerInventory) {
  this(containerId, playerInventory, ContainerLevelAccess.NULL);
}

// Server menu constructor
public MyMenuAccess(int containerId, Inventory playerInventory, ContainerLevelAccess access) {
  // ...
}
```

### Data Synchronization

Some data needs to be present on both the server and the client to display to the player. To do this, the menu implements a basic layer of data synchronization such that whenever the current data does not match the data last synced to the client. For players, this is checked every tick.

Minecraft supports two forms of data synchronization by default: `ItemStack`s via `Slot`s and integers via `DataSlot`s. `Slot`s and `DataSlot`s are views which hold references to data storages that can be be modified by the player in a screen, assuming the action is valid. These can be added to a menu within the constructor through `#addSlot` and `#addDataSlot`.

!!! note

>     Since `Container`s used by `Slot`s are deprecated by Forge in favor of using the [`IItemHandler` capability][cap], the rest of the explanation will revolve around using `SlotItemHandler`: the capability variant.

Reviewer: The nouns around that colon should be switched around.
```suggestion
    Since `Container`s used by `Slot`s are deprecated by Forge in favor of using the [`IItemHandler` capability][cap], the rest of the explanation will revolve around using the capability variant: `SlotItemHandler`.
```

A `SlotItemHandler` contains four parameters: the `IItemHandler` representing the inventory the stacks are within, the index of the stack this slot is specifically representing, and the x and y position of where the top-left position of the slot will render on the screen relative to `AbstractContainerScreen#leftPos` and `#topPos`. The client menu constructor should always supply an empty instance of an inventory of the same size.

In most cases, any slots the menu contains is first added, followed by the player's inventory, and finally concluded with the player's hotbar. To access any individual `Slot` from the menu, the index must be calculated based upon the order of which slots were added.

> A `DataSlot` is an abstract class which should implement a getter and setter to reference the data stored in the data storage object.The client menu constructor should always supply a new instance via `DataSlot#standalone`.

Reviewer: Missing space.
```suggestion
A `DataSlot` is an abstract class which should implement a getter and setter to reference the data stored in the data storage object. The client menu constructor should always supply a new instance via `DataSlot#standalone`.
```

These, along with slots, should be recreated every time a new menu is initialized.

> !!! warning
>     Although a `DataSlot` may hold enough data for an integer, only a short is sent across the network.

Reviewer: Rephrased it a bit with a bit more emphasis to draw the eye. Also, what happens when a value larger than a short is provided in a DataSlot? Is it truncated to fit the bits of a short, or can some form of overflow/underflow occur?

```suggestion
!!! warning
    Although a `DataSlot` stores an integer, it is effectively limited to a **short** (-32768 to 32767) because of how it sends the value across the network.
```

```java
// Assume we have an inventory from a data object of size 5
// Assume we have a DataSlot constructed on each initialization of the server menu

// Client menu constructor
public MyMenuAccess(int containerId, Inventory playerInventory) {
  this(containerId, playerInventory, new ItemStackHandler(5), DataSlot.standalone());
}

// Server menu constructor
public MyMenuAccess(int containerId, Inventory playerInventory, IItemHandler dataInventory, DataSlot dataSingle) {
  // Check if the data inventory size is some fixed value
  // Then, add slots for data inventory
  this.addSlot(new SlotItemHandler(dataInventory, /*...*/));

  // Add slots for player inventory
  this.addSlot(new Slot(playerInventory, /*...*/));

  // Add data slots for handled integers
  this.addDataSlot(dataSingle);

  // ...
}
```

#### `ContainerData`

> If multiple integers need to be synced to the client, a `ContainerData` can be used to reference the integers instead. This interface functions as an index lookup such that each index represents a different integer. `ContainerData`s can also be constructed in the data object itself if the `ContainerData` is added to the menu through `#addDataSlots`. The method creates a new `DataSlot` for the amount of data specified by the interface. The client menu constructor should always supply a new instance via `SimpleContainerData`.

Reviewer: These also fall to the same caveat as `DataSlot`s with regards to the effectively `short` data size, right?

// We also have the 27 player inventory slots and the 9 hotbar slots
// As such, the actual slots are indexed like so:
//   - Data Inventory: Result (0), Inputs (1 - 4)
//   - Player Inventory (5 - 31)
//   - Player Hotbar (32 - 40)
@Override
public ItemStack quickMoveStack(Player player, int quickMovedSlotIndex) {
  // The quick moved slot stack
  ItemStack quickMovedStack = ItemStack.EMPTY;
  // The quick moved slot
  Slot quickMovedSlot = this.slots.get(quickMovedSlotIndex) 
  
   // If the slot is in the valid range and the slot is not empty
  if (quickMovedSlot != null && quickMovedSlot.hasItem()) {
    // Get the raw stack to move
    ItemStack rawStack = quickMovedSlot.getItem(); 
    // Set the slot stack to a copy of the raw stack
    quickMovedStack = rawStack.copy();

    /*
    The following quick move logic can be simplified to if in data inventory,
    try to move to player inventory/hotbar and vice versa for containers
    that cannot transform data (e.g. chests).
    */

    // If the quick move was performed on the data inventory result slot
    if (quickMovedSlotIndex == 0) {
      // Try to move the result slot into the player inventory/hotbar

>       if (!this.moveItemStackTo(rawStack, 5, 41, true)) {

Reviewer: `moveItemStackTo`'s slot end range is exclusive?

}
```

!!! warning
    As `ContainerData` delegates to `DataSlot`s, these are also limited to a **short** (-32768 to 32767).

#### `#quickMoveStack`

`#quickMoveStack` is the second method that must be implemented by any menu. This method is called whenever a stack has been shift-clicked, or quick moved, out of its current slot until the stack has been fully moved out of its previous slot or there is no other place for the stack to go. The method returns a copy of the stack in the slot being quick moved.

> Stacks are typically moved between slots using `#moveItemStackTo`, which moves the stack into the first available slot. It takes in the stack to be moved, the starting slot index (inclusive) to try and move the stack to, the last slot index (exclusive), and whether to check the slots in ascending (when false) or descending (when true) order.

Reviewer: Just to be clear, ascending means from the starting index to the ending index, and descending means from the ending index to the starting index? 

Also, there should be better parallelism: either "starting slot index" and "ending slot index", or "first slot index" and "last slot index".

As a matter of style, should we always monospace `true` and `false`? If not, when do we monospace them?

Author: I try to always monospace them, just forgot this time.
